File: leetcode/cpp/530. 二叉搜索树的最小绝对差.cpp
typedef struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
} TreeNode;

class Solution {
public:
    // 中序遍历，更新差值
    void mid(TreeNode* root, int& ret, int& lastVal) {
        if (root == nullptr) {
            return;
        }
        mid(root->left, ret, lastVal);
        if (root->val - lastVal < ret) {
            ret = root->val - lastVal;
        }
        lastVal = root->val;
        mid(root->right, ret, lastVal);
    }

    int getMinimum(TreeNode* root) {
        int ret = 0x7fffffff;
        int lastVal = 0x80000000;
        mid(root, ret, lastVal);
        return ret;
    }
};

File: leetcode/python/530. 二叉搜索树的最小绝对差.py
#!/bin/bash python
"""
530. 二叉搜索树的最小绝对差

思路：
中序遍历，按照有序数组直接的差值来更新
"""

def Solution(obejct):
    def getMinimumDifference(self, root):
        self.ret = float("+inf")
        self.lastVal = float("-inf")
        def mid(root):
            if root == None:
                return
            mid(root.left)
            self.ret = min(self.ret, root.val - self.lastVal)
            self.lastVal = root.val
            mid(root.right)
            return
        mid(root)
        return self.ret